### PR TITLE
UPM package

### PR DESCRIPTION
**Pull request by Arshia001**

### CODE REVIEW

**Review by juans-chainsafe — CHANGES_REQUESTED** [review]
@Arshia001 WebWallet looks good so far, I need to do more testing.

In WebGL I'm having an error with the ERC20 prefabs in the SampleScene, after I do the login, I get this error on the console:
```
webgl.framework.js.br:3 JsonException: No parameterless constructor defined for 'Nethereum.JsonRpc.Client.RpcMessages.RpcParametersJsonConverter'.
  at Newtonsoft.Json.Serialization.JsonTypeReflector+<>c__DisplayClass22_0.<GetCreator>b__0 (System.Object[] parameters) [0x00000] in <00000000000000000000000000000000>:0 
Rethrow as JsonException: Error creating 'Nethereum.JsonRpc.Client.RpcMessages.RpcParametersJsonConverter'.
  at Newtonsoft.Json.Serialization.JsonTypeReflector+<>c__DisplayClass22_0.<GetCreator>b__0 (System.Object[] parameters) [0x00000] in <00000000000000000000000000000000>:0 
--- End of stack trace from previous location where exception was thrown ---
```

I tried with other prefabs like SignMessage or Sha3, and they didn't fail on WebGL

**Comment by KBryan** [comment]
Hey, @juans-chainsafe can you try adding the following to the assets folder? Name the file link.xml

```
<linker>
	<assembly fullname="System" preserve="all">
		<type fullname="System.ComponentModel.TypeConverter" preserve="all"/>
		<type fullname="System.ComponentModel.ArrayConverter" preserve="all"/>
		<type fullname="System.ComponentModel.BaseNumberConverter" preserve="all"/>
		<type fullname="System.ComponentModel.BooleanConverter" preserve="all"/>
		<type fullname="System.ComponentModel.ByteConverter" preserve="all"/>
		<type fullname="System.ComponentModel.CharConverter" preserve="all"/>
		<type fullname="System.ComponentModel.CollectionConverter" preserve="all"/>
		<type fullname="System.ComponentModel.ComponentConverter" preserve="all"/>
		<type fullname="System.ComponentModel.CultureInfoConverter" preserve="all"/>
		<type fullname="System.ComponentModel.DateTimeConverter" preserve="all"/>
		<type fullname="System.ComponentModel.DecimalConverter" preserve="all"/>
		<type fullname="System.ComponentModel.DoubleConverter" preserve="all"/>
		<type fullname="System.ComponentModel.EnumConverter" preserve="all"/>
		<type fullname="System.ComponentModel.ExpandableObjectConverter" preserve="all"/>
		<type fullname="System.ComponentModel.Int16Converter" preserve="all"/>
		<type fullname="System.ComponentModel.Int32Converter" preserve="all"/>
		<type fullname="System.ComponentModel.Int64Converter" preserve="all"/>
		<type fullname="System.ComponentModel.NullableConverter" preserve="all"/>
		<type fullname="System.ComponentModel.SByteConverter" preserve="all"/>
		<type fullname="System.ComponentModel.SingleConverter" preserve="all"/>
		<type fullname="System.ComponentModel.StringConverter" preserve="all"/>
		<type fullname="System.ComponentModel.TimeSpanConverter" preserve="all"/>
		<type fullname="System.ComponentModel.UInt16Converter" preserve="all"/>
		<type fullname="System.ComponentModel.UInt32Converter" preserve="all"/>
		<type fullname="System.ComponentModel.UInt64Converter" preserve="all"/>
	</assembly>
	<assembly fullname="Nethereum.ABI" preserve="all"/>
	<assembly fullname="Nethereum.Hex" preserve="all"/>
	<assembly fullname="Nethereum.RPC" preserve="all"/>
	<assembly fullname="Nethereum.RLP" preserve="all"/>
	<assembly fullname="Nethereum.Contracts" preserve="all"/>
	<assembly fullname="Nethereum.Util" preserve="all"/>
	<assembly fullname="Nethereum.Signer" preserve="all"/>
	<assembly fullname="Nethereum.JsonRpc.Client" preserve="all"/>
	<assembly fullname="Nethereum.Unity" preserve="all"/>
</linker>
```

**Comment by juans-chainsafe** [comment]
> Hey, @juans-chainsafe can you try adding the following to the assets folder? Name the file link.xml
> 
> ```
> <linker>
> 	<assembly fullname="System" preserve="all">
> 		<type fullname="System.ComponentModel.TypeConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.ArrayConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.BaseNumberConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.BooleanConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.ByteConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.CharConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.CollectionConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.ComponentConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.CultureInfoConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.DateTimeConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.DecimalConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.DoubleConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.EnumConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.ExpandableObjectConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.Int16Converter" preserve="all"/>
> 		<type fullname="System.ComponentModel.Int32Converter" preserve="all"/>
> 		<type fullname="System.ComponentModel.Int64Converter" preserve="all"/>
> 		<type fullname="System.ComponentModel.NullableConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.SByteConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.SingleConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.StringConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.TimeSpanConverter" preserve="all"/>
> 		<type fullname="System.ComponentModel.UInt16Converter" preserve="all"/>
> 		<type fullname="System.ComponentModel.UInt32Converter" preserve="all"/>
> 		<type fullname="System.ComponentModel.UInt64Converter" preserve="all"/>
> 	</assembly>
> 	<assembly fullname="Nethereum.ABI" preserve="all"/>
> 	<assembly fullname="Nethereum.Hex" preserve="all"/>
> 	<assembly fullname="Nethereum.RPC" preserve="all"/>
> 	<assembly fullname="Nethereum.RLP" preserve="all"/>
> 	<assembly fullname="Nethereum.Contracts" preserve="all"/>
> 	<assembly fullname="Nethereum.Util" preserve="all"/>
> 	<assembly fullname="Nethereum.Signer" preserve="all"/>
> 	<assembly fullname="Nethereum.JsonRpc.Client" preserve="all"/>
> 	<assembly fullname="Nethereum.Unity" preserve="all"/>
> </linker>
> ```

I get the following build error after adding the file: 
Screenshot 2023-06-01 at 15 22 00

**Comment by juans-chainsafe** [comment]
> Okay, also try adding this in Unity preferences. Strip Enginee Code unckecked Managed Stripping Level low
> 
> managed-code-stripping-level

I still have the same errors, I had Strip Enginee Code checked and stripping level on low (I put it on minimum)
Screenshot 2023-06-01 at 16 12 27

I get the errors when I try to generate the build

**Comment by juans-chainsafe** [comment]
> Strip Enginee

@Arshia001 UPDATE: after some digging with @KBryan we removed the xml file that we added, and tried another WebGL build setting:

-  Strip Enginee Code unchecked
-  Stripping level on minimum 
And it worked! I can continue with the testing now.

We will need to do this every time that we import the project... so it will be nice if it can be fixed automatically.

**Review by juans-chainsafe — APPROVED** [review]
Didn't find more errors, only a few things that we need to take a look:

@sneakzttv @Arshia001 the Web3Auth modal looks weird, not sure what happened with the styles there, but it works.
Screenshot 2023-06-02 at 11 29 36

@Arshia001 looks like (and it makes sense) that if I add prefabs into the SampleScene, I get the src/UnitySampleProject/Assets/Scenes/SampleScene.Unity updated in github, not sure if something can be done to avoid this, probably not.
Screenshot 2023-06-02 at 11 31 36

Also, when I change the build settings, these files are updated:

- src/UnitySampleProject/ProjectSettings/EditorBuildSettings.asset
- src/UnitySampleProject/ProjectSettings/ProjectSettings.asset

**Comment by KBryan** [comment]
@juans-chainsafe is the sample project needed? If not it doesn't need to be shipped out we should remove it and the associated metadata.
@sneakzttv for the Web3WalletUI we may need to update this to Unity's UIKit.

**Comment by Arshia001** [comment]
@juans-chainsafe I removed the textures of the wallet UI instead of moving them, that one's one me.

regarding `SampleScene` being updated, there's no way to avoid that, we just have to make sure we don't commit it.

@KBryan the sample is going to act as both a place for us to test the code (hopefully automatically and in the CI pipeline), as well as a sample for users on how to set everything up and a place where they can get their hands dirty with the SDK.

**Comment by KBryan** [comment]
gotcha sounds good.

File: src/UnityPackages/io.chainsafe.web3-unity/Runtime/Scripts/ProjectConfigUtilities.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+public static class ProjectConfigUtilities
+{
+    public static ProjectConfigScriptableObject Load()
+    {
+        var projectConfig = Resources.Load<ProjectConfigScriptableObject>("ProjectConfigData");
+        return projectConfig ? projectConfig : null;
+    }
+
+#if UNITY_EDITOR
```
Comment (oleksandrchainsafe) [inline src/UnityPackages/io.chainsafe.web3-unity/Runtime/Scripts/ProjectConfigUtilities.cs:13]: Oh snap, I've added some similar logic at #453. We'd have to remove the duplicating logic from my side.

File: src/ChainSafe.GamingSdk.EVM.Unity/publish-to-unity-package.bat
[the remark this answers is not in the document]
Comment (Arshia001) [reply src/ChainSafe.GamingSdk.EVM.Unity/publish-to-unity-package.bat:1]: They don't; they end up in the UPM package, which is pulled into the sample project by a path reference.

File: src/UnityPackages/io.chainsafe.web3-unity/Editor/WebGLTemplateSync.cs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        private static void CopyFolder(DirectoryInfo source, DirectoryInfo destination)
+        {
+            if (!destination.Exists)
+            {
+                destination.Create();
+            }
+
+            foreach (var file in source.GetFiles())
+            {
+                if (file.Extension != ".meta")
```
[the remark this answers is not in the document]
Comment (Arshia001) [reply src/UnityPackages/io.chainsafe.web3-unity/Editor/WebGLTemplateSync.cs:81]: There's even an IDE code fix for that! 😄 good catch.

File: src/UnityPackages/io.chainsafe.web3-unity/Runtime/Scripts/ProjectConfigUtilities.cs (hunk 1 above)
Comment (Arshia001) [reply src/UnityPackages/io.chainsafe.web3-unity/Runtime/Scripts/ProjectConfigUtilities.cs:13]: yep.

**Review by oleksandrchainsafe — DISMISSED** [review]

**Review by juans-chainsafe — CHANGES_REQUESTED** [review]
@Arshia001 WebGL and WebWallet builds are working correctly, but:
I'm getting this error when I try to save the ChainSafeServerSettings (they are saved but the code for writing the networks.js file is failing):
Screenshot 2023-06-07 at 12 29 51

```
CreateAsset() should not be used to create a file of type 'js' - consider using AssetDatabase.ImportAsset() to import an existing 'js' file instead or change the file type to '*.asset'. This error will in a future release be changed to an exception.
UnityEngine.StackTraceUtility:ExtractStackTrace ()
ChainSafeServerSettings:WriteNetworkFile () (at /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/src/UnityPackages/io.chainsafe.web3-unity/Editor/ServerSettings.cs:200)
ChainSafeServerSettings/<ValidateProjectID>d__17:MoveNext () (at /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/src/UnityPackages/io.chainsafe.web3-unity/Editor/ServerSettings.cs:139)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<bool>:SetResult (bool)
ChainSafeServerSettings/<ValidateProjectIDAsync>d__18:MoveNext () (at /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/src/UnityPackages/io.chainsafe.web3-unity/Editor/ServerSettings.cs:176)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder:SetResult ()
EditorUtilities/<SendAndWait>d__0:MoveNext () (at /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/src/UnityPackages/io.chainsafe.web3-unity/Editor/EditorUtilities.cs:18)
UnityEngine.UnitySynchronizationContext:ExecuteTasks ()
```

**Comment by juans-chainsafe** [comment]
> @juans-chainsafe should be fixed now. Unity does some very odd things indeed.

Now is not failing, but I don't see the networks.js updated, it has the defaut values:
```
window.networks = [
  {
    id: 5,
    label: "Ethereum Goerli",
    token: "goETH",
    rpcUrl: `https://goerli.infura.io/v3/assdfsdf`,
  }
]
```
Also, in the console logs I dont see any more messages after the "Updating network.js", at least now is saving the symbol
Screenshot 2023-06-07 at 14 13 53

**Comment by Arshia001** [comment]
@juans-chainsafe Should be working now. I also added one more log message so it doesn't look like it got stuck.

**Comment by juans-chainsafe** [comment]
> @juans-chainsafe Should be working now. I also added one more log message so it doesn't look like it got stuck.

I see the new log that says "Done" but the network.js file is not updated correctly, now I see this:
```
window.networks = [
  {
    id: 5,
    label: "ethereum goerli",
    token: "Please enter your chain's native symbol i.e Eth, Cro",
    rpcUrl: 'https://goerli.infura.io/v3/904006115c764661965dc0909e5ed473',
  }
]
```
But I set the Symbol in the settings on Unity, but I don't see it updated on the network.js
Screenshot 2023-06-08 at 10 57 05

**Comment by sneakzttv** [comment]
i'll see if i can find it

**Comment by sneakzttv** [comment]
working 
image

**Review by sneakzttv — APPROVED** [review]

**Review by juans-chainsafe — APPROVED** [review]
Working now! Symbol is being saved to network.js

**Pull request merged**